spectrum-kes/src/prod_comp.rs

```diff
@@ -0,0 +1,296 @@
+
+use ecdsa::hazmat::SignPrimitive;
+use ecdsa::signature::{Signer, SignerMut, Verifier};
+use ecdsa::{Signature, SigningKey, VerifyingKey};
+use elliptic_curve::generic_array::ArrayLength;
+use elliptic_curve::point::PointCompression;
+use elliptic_curve::sec1::{FromEncodedPoint, ModulusSize, ToEncodedPoint};
+use elliptic_curve::{CurveArithmetic, PublicKey, SecretKey};
+
+use spectrum_crypto::digest::Sha2Digest256;
+use spectrum_vrf::utils::projective_point_to_bytes;
+
+use crate::utils::{associate_message_with_slot, associate_pk_with_slot, double_the_seed, kes_key_gen};
+
+#[derive(Debug)]
+pub struct Error;
+
+pub struct KesProdSecretKey<TCurve: CurveArithmetic + ecdsa::PrimeCurve> {
+    sk_0: SecretKey<TCurve>,
+    sk_1: SecretKey<TCurve>,
+    sig: Signature<TCurve>,
+    pk_1: PublicKey<TCurve>,
+    seed_11: Sha2Digest256,
+}
+
+pub struct KesProdSignature<TCurve: CurveArithmetic + ecdsa::PrimeCurve> {
+    pk_1: PublicKey<TCurve>,
+    sig_0: Signature<TCurve>,
+    sig_1: Signature<TCurve>,
+}
+
+pub fn kes_prod_key_gen<TCurve: CurveArithmetic + ecdsa::PrimeCurve + PointCompression>(
+    seed: &Sha2Digest256,
+) -> Result<(KesProdSecretKey<TCurve>, PublicKey<TCurve>), Error>
+where
+    <TCurve as CurveArithmetic>::AffinePoint: FromEncodedPoint<TCurve>,
+    <TCurve as elliptic_curve::Curve>::FieldBytesSize: ModulusSize,
+    <TCurve as CurveArithmetic>::AffinePoint: ToEncodedPoint<TCurve>,
+    <TCurve as CurveArithmetic>::Scalar: SignPrimitive<TCurve>,
+    <<TCurve as elliptic_curve::Curve>::FieldBytesSize as Add>::Output: ArrayLength<u8>,
+    SigningKey<TCurve>: Signer<Signature<TCurve>>,
+    SigningKey<TCurve>: SignerMut<Signature<TCurve>>,
+{
+    let (seed_0, seed_01) = double_the_seed(&seed);
+    let (seed_10, seed_11) = double_the_seed(&seed_01);
+
+    let (sk_0, pk) = kes_key_gen::<TCurve>(&seed_0).unwrap();
+    let (sk_1, pk_1) = kes_key_gen::<TCurve>(&seed_10).unwrap();
+
+    let signing_key = SigningKey::from(&sk_0);
+    let sig = signing_key.sign(&projective_point_to_bytes::<TCurve>(pk_1.to_projective()));
+
+    Ok((
+        KesProdSecretKey {
+            sk_0,
+            sk_1,
+            sig,
+            pk_1,
+            seed_11,
+        },
+        pk,
+    ))
+}
+
+fn kes_prod_update<TCurve: CurveArithmetic + ecdsa::PrimeCurve + PointCompression>(
+    kes_sk: &KesProdSecretKey<TCurve>,
+    current_slot: &i32,
+    bound_slot: &i32,
+) -> Result<KesProdSecretKey<TCurve>, Error>
+where
+    <TCurve as CurveArithmetic>::AffinePoint: FromEncodedPoint<TCurve>,
+    <TCurve as elliptic_curve::Curve>::FieldBytesSize: ModulusSize,
+    <TCurve as CurveArithmetic>::AffinePoint: ToEncodedPoint<TCurve>,
+    <TCurve as CurveArithmetic>::Scalar: SignPrimitive<TCurve>,
+    <<TCurve as elliptic_curve::Curve>::FieldBytesSize as Add>::Output: ArrayLength<u8>,
+    SigningKey<TCurve>: Signer<Signature<TCurve>>,
+    SigningKey<TCurve>: SignerMut<Signature<TCurve>>,
+{
+    //TODO: make recursive
```

Comment:
What about this?

Reply:
Unfinished yet, I interrupted to edit the Wp, will continue from tomorrow